controllers/vspherecluster_controller.go

```diff
@@ -290,56 +455,118 @@ func (r clusterReconciler) reconcileControlPlaneEndpoint(ctx *context.ClusterCon
+	apiServerTriggers   = map[types.UID]struct{}{}
+	apiServerTriggersMu sync.Mutex
+)
+
+func (r clusterReconciler) reconcileVSphereClusterWhenAPIServerIsOnline(ctx *context.ClusterContext) {
+	if ctx.Cluster.Status.ControlPlaneInitialized {
+		ctx.Logger.Info("skipping reconcile when API server is online",
+			"reason", "controlPlaneInitialized")
+		return
+	}
+	apiServerTriggersMu.Lock()
+	defer apiServerTriggersMu.Unlock()
+	if _, ok := apiServerTriggers[ctx.Cluster.UID]; ok {
+		ctx.Logger.Info("skipping reconcile when API server is online",
+			"reason", "alreadyPolling")
+		return
+	}
+	apiServerTriggers[ctx.Cluster.UID] = struct{}{}
+	go func() {
+		// Block until the target API server is online.
+		ctx.Logger.Info("start polling API server for online check")
+		wait.PollImmediateInfinite(time.Second*1, func() (bool, error) { return r.isAPIServerOnline(ctx), nil })
+		ctx.Logger.Info("stop polling API server for online check")
+		ctx.Logger.Info("triggering GenericEvent", "reason", "api-server-online")
+		eventChannel := ctx.GetGenericEventChannelFor(ctx.VSphereCluster.GetObjectKind().GroupVersionKind())
+		eventChannel <- event.GenericEvent{
+			Meta:   ctx.VSphereCluster,
+			Object: ctx.VSphereCluster,
+		}
+
+		// Once the control plane has been marked as initialized it is safe to
+		// remove the key from the map that prevents multiple goroutines from
+		// polling the API server to see if it is online.
+		ctx.Logger.Info("start polling for control plane initialized")
+		wait.PollImmediateInfinite(time.Second*1, func() (bool, error) { return r.isControlPlaneInitialized(ctx), nil })
+		ctx.Logger.Info("stop polling for control plane initialized")
+		apiServerTriggersMu.Lock()
+		delete(apiServerTriggers, ctx.Cluster.UID)
+		apiServerTriggersMu.Unlock()
+	}()
 }
 
 func (r clusterReconciler) isAPIServerOnline(ctx *context.ClusterContext) bool {
-	if client, err := infrautilv1.NewKubeClient(ctx, ctx.Client, ctx.Cluster); err == nil {
-		if _, err := client.CoreV1().Nodes().List(metav1.ListOptions{}); err == nil {
+	if kubeClient, err := infrautilv1.NewKubeClient(ctx, ctx.Client, ctx.Cluster); err == nil {
+		if _, err := kubeClient.CoreV1().Nodes().List(metav1.ListOptions{}); err == nil {
+			// The target cluster is online. To make sure the correct control
+			// plane endpoint information is logged, it is necessary to fetch
+			// an up-to-date Cluster resource. If this fails, then set the
+			// control plane endpoint information to the values from the
+			// VSphereCluster resource, as it must have the correct information
+			// if the API server is online.
+			cluster := &clusterv1.Cluster{}
+			clusterKey := client.ObjectKey{Namespace: ctx.Cluster.Namespace, Name: ctx.Cluster.Name}
+			if err := ctx.Client.Get(ctx, clusterKey, cluster); err != nil {
+				cluster = ctx.Cluster.DeepCopy()
+				cluster.Spec.ControlPlaneEndpoint.Host = ctx.VSphereCluster.Spec.ControlPlaneEndpoint.Host
```

Comment:
why are we setting these ? as cluster is local, maybe ctx.Cluster ?

Reply:
Hi @yastij,

Strictly for the logging statement below.